pkg/operator/controller/externaldns/credentials-request_test.go

```diff
@@ -0,0 +1,287 @@
+				codec, _ := cco.NewCodec()
+				azureSpec := cco.AzureProviderSpec{}
+				err := codec.DecodeProviderSpec(got.Spec.ProviderSpec, &azureSpec)
+				if err != nil {
+					t.Errorf("error decoding provider v1 spec: %v", err)
+				}
+
+				expectedAzureSpec := cco.AzureProviderSpec{}
+				err = codec.DecodeProviderSpec(tc.expectedCredentialRequest.Spec.ProviderSpec, &expectedAzureSpec)
+				if err != nil {
+					t.Errorf("error decoding provider v1 spec: %v", err)
+				}
+
+				if reflect.DeepEqual(azureSpec, expectedAzureSpec) == false {
+					t.Errorf("Got CredentialRequest Spec %v\n  but expected CredentialRequest Spec is %v", azureSpec, expectedAzureSpec)
+				}
+			}
+
+			if tc.inputExtDNS.Spec.Provider.Type == operatorv1alpha1.ProviderTypeGCP {
+				codec, _ := cco.NewCodec()
+				gcpSpec := cco.GCPProviderSpec{}
+				err := codec.DecodeProviderSpec(got.Spec.ProviderSpec, &gcpSpec)
+				if err != nil {
+					t.Errorf("error decoding provider v1 spec: %v", err)
+				}
+
+				expectedGCPSpec := cco.GCPProviderSpec{}
+				err = codec.DecodeProviderSpec(tc.expectedCredentialRequest.Spec.ProviderSpec, &expectedGCPSpec)
+				if err != nil {
+					t.Errorf("error decoding provider v1 spec: %v", err)
+				}
+
+				if reflect.DeepEqual(gcpSpec, expectedGCPSpec) == false {
+					t.Errorf("Got CredentialRequest Spec %v\n  but expected CredentialRequest Spec is %v", gcpSpec, expectedGCPSpec)
+				}
+			}
+
+		})
+
+	}
+
+}
+
+func testGCPExtDNSInstanceWhenSourceIsOCPRoute() *operatorv1alpha1.ExternalDNS {
+	extDNS := testExtDNSInstanceforOCPRouteSource()
+	extDNS.Spec.Provider = operatorv1alpha1.ExternalDNSProvider{
+		Type: operatorv1alpha1.ProviderTypeGCP,
+		GCP:  &operatorv1alpha1.ExternalDNSGCPProviderOptions{},
+	}
+	return extDNS
+}
+
+func testAzureExtDNSInstanceWhenSourceIsOCPRoute() *operatorv1alpha1.ExternalDNS {
+	extDNS := testExtDNSInstanceforOCPRouteSource()
+	extDNS.Spec.Provider = operatorv1alpha1.ExternalDNSProvider{
+		Type:  operatorv1alpha1.ProviderTypeAzure,
+		Azure: &operatorv1alpha1.ExternalDNSAzureProviderOptions{},
+	}
+	return extDNS
+}
+
+func getCredentialRequest(name types.NamespacedName, providerType operatorv1alpha1.ExternalDNSProviderType) *cco.CredentialsRequest {
+
+	credentialsRequest := &cco.CredentialsRequest{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "CredentialsRequest",
+			APIVersion: "cloudcredential.openshift.io/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name.Name,
+			Namespace: name.Namespace,
+		},
+		Spec: cco.CredentialsRequestSpec{
+			ServiceAccountNames: []string{"external-dns-operator"},
+			SecretRef: k8sv1.ObjectReference{
+				Name:      "externaldns-cloud-credentials",
+				Namespace: "external-dns-operator",
+			},
+		},
+	}
+
+	switch providerType {
+
+	case operatorv1alpha1.ProviderTypeAWS:
+		codec, err := cco.NewCodec()
+		if err != nil {
+			fmt.Printf("error creating new codec: %v", err)
+			return nil
+		}
+		awsProviderSpec, err := awsProviderConfig(codec)
```

Comment:
I don't quite get this. `ensureExternalCredentialsRequest` uses `<provider>ProviderConfig()` functions. Then you use the same function to get the expected object's provider spec. Obviously it will always be the same. What is tested then?

Reply:
we have to decode for comparison or else it tries to compares addresses and the cmp or deep reflect fails as in expected a "\n" char is getting added

Reply:
What I meant was not about the encoding or decoding. I wanted to say that the spec you generate as the expected value will always be the same as the one generated by the function you are testing. This makes the test pointless.

Reply:
no it does not make the test pointless as we are checking the contents of the provider spec if they match.

Reply:
As mentioned to you in the expect a new line gets added not sure why. So after decoding we see human readable output and the reflect equal and go cmp return it as true.

Reply:
you can try what I am saying...just try to compare the encoded the values